### PR TITLE
New Resource: data_factory_integration_runtime_managed

### PR DESCRIPTION
**Pull request by r0bnet**
Requested and discussed here: https://github.com/terraform-providers/terraform-provider-azurerm/issues/3275

### CODE REVIEW

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+		if numberOfNodes := computeProps.NumberOfNodes; numberOfNodes != nil {
+			d.Set("number_of_nodes", numberOfNodes)
+		}
+
+		if maxParallelExecutionsPerNode := computeProps.MaxParallelExecutionsPerNode; maxParallelExecutionsPerNode != nil {
+			d.Set("max_parallel_executions_per_node", maxParallelExecutionsPerNode)
+		}
+
+		if vnetProps := computeProps.VNetProperties; vnetProps != nil {
+			d.Set("vnet_integration", flattenArmDataFactoryIntegrationRuntimeManagedVnetIntegration(vnetProps))
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed.go]: We should check for an error here
```suggestion
			if err := d.Set("vnet_integration", flattenArmDataFactoryIntegrationRuntimeManagedVnetIntegration(vnetProps)); err != nil {
			     return fmt.Error("Error setting `vnet_integration`: %+v", err)
			}
```

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+		}
+	}
+
+	if ssisProps := managedIntegrationRuntime.SsisProperties; ssisProps != nil {
+		d.Set("edition", string(ssisProps.Edition))
+		d.Set("license_type", string(ssisProps.LicenseType))
+
+		if catalogInfoProps := ssisProps.CatalogInfo; catalogInfoProps != nil {
+			d.Set("catalog_info", flattenArmDataFactoryIntegrationRuntimeManagedSsisCatalogInfo(catalogInfoProps, d))
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed.go]: We should check for an error while setting.

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+		}
+
+		if customSetupScriptProps := ssisProps.CustomSetupScriptProperties; customSetupScriptProps != nil {
+			d.Set("custom_setup_script", flattenArmDataFactoryIntegrationRuntimeManagedSsisCustomSetupScript(customSetupScriptProps, d))
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed.go]: And here

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+	location := azure.NormalizeLocation(d.Get("location").(string))
+	computeProperties := datafactory.IntegrationRuntimeComputeProperties{
+		Location:                     &location,
+		NodeSize:                     utils.String(d.Get("node_size").(string)),
+		NumberOfNodes:                utils.Int32(int32(d.Get("number_of_nodes").(int))),
+		MaxParallelExecutionsPerNode: utils.Int32(int32(d.Get("max_parallel_executions_per_node").(int))),
+	}
+
+	if _, ok := d.GetOk("vnet_integration"); ok {
+		vnetProps := d.Get("vnet_integration").([]interface{})[0].(map[string]interface{})
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed.go]: `d.Get("vnet_integration").([]interface{})[0]` has the potential to be nil so we should check before casting it to `map[string]interface`

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+}
+
+func expandArmDataFactoryIntegrationRuntimeManagedSsisProperties(d *schema.ResourceData) *datafactory.IntegrationRuntimeSsisProperties {
+	ssisProperties := &datafactory.IntegrationRuntimeSsisProperties{
+		Edition:     datafactory.IntegrationRuntimeEdition(d.Get("edition").(string)),
+		LicenseType: datafactory.IntegrationRuntimeLicenseType(d.Get("license_type").(string)),
+	}
+
+	if _, ok := d.GetOk("catalog_info"); ok {
+		catalogInfo := d.Get("catalog_info").([]interface{})[0].(map[string]interface{})
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed.go]: And here we should confirm that `d.Get("catalog_info").([]interface{})[0]` isn't nil

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+		ssisProperties.CatalogInfo = &datafactory.IntegrationRuntimeSsisCatalogInfo{
+			CatalogServerEndpoint: utils.String(catalogInfo["server_endpoint"].(string)),
+			CatalogAdminUserName:  utils.String(catalogInfo["administrator_login"].(string)),
+			CatalogAdminPassword:  adminPassword,
+			CatalogPricingTier:    datafactory.IntegrationRuntimeSsisCatalogPricingTier(catalogInfo["pricing_tier"].(string)),
+		}
+	}
+
+	if _, ok := d.GetOk("custom_setup_script"); ok {
+		customSetupScript := d.Get("custom_setup_script").([]interface{})[0].(map[string]interface{})
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed.go]: And here we should confirm that `d.Get("custom_setup_script").([]interface{})[0]` isn't nil

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+
+		if numberOfNodes := computeProps.NumberOfNodes; numberOfNodes != nil {
+			d.Set("number_of_nodes", numberOfNodes)
+		}
+
+		if maxParallelExecutionsPerNode := computeProps.MaxParallelExecutionsPerNode; maxParallelExecutionsPerNode != nil {
+			d.Set("max_parallel_executions_per_node", maxParallelExecutionsPerNode)
+		}
+
+		if vnetProps := computeProps.VNetProperties; vnetProps != nil {
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed.go]: We always want to set a value (ie. `vnet_integration` but if `vnetProps` is nil then we'll never set `vnet_integration`. We should put the nil check logic in `flattenArmDataFactoryIntegrationRuntimeManagedVnetIntegration` and return an empty `[]interface{}` if it's nil so `vnet_integration` is set.

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+			d.Set("vnet_integration", flattenArmDataFactoryIntegrationRuntimeManagedVnetIntegration(vnetProps))
+		}
+	}
+
+	if ssisProps := managedIntegrationRuntime.SsisProperties; ssisProps != nil {
+		d.Set("edition", string(ssisProps.Edition))
+		d.Set("license_type", string(ssisProps.LicenseType))
+
+		if catalogInfoProps := ssisProps.CatalogInfo; catalogInfoProps != nil {
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed.go]: We always want to set a value so similar to the above comment, we should move the nil logic into `flattenArmDataFactoryIntegrationRuntimeManagedSsisCatalogInfo` and return an empty `[]interface{}` if `catalogInfoProps` is nil.

File: azurerm/resource_arm_data_factory_integration_runtime_managed_test.go [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMDataFactoryIntegrationRuntimeManagedExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "vnet_integration.#", "1"),
+					resource.TestCheckResourceAttrSet(resourceName, "vnet_integration.0.vnet_id"),
+					resource.TestCheckResourceAttrSet(resourceName, "vnet_integration.0.subnet_name"),
+				),
+			},
+		},
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed_test.go]: We should add an import test step here

File: azurerm/resource_arm_data_factory_integration_runtime_managed_test.go [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMDataFactoryIntegrationRuntimeManagedExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "custom_setup_script.#", "1"),
+					resource.TestCheckResourceAttrSet(resourceName, "custom_setup_script.0.blob_container_uri"),
+					resource.TestCheckResourceAttrSet(resourceName, "custom_setup_script.0.sas_token"),
+				),
+			},
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed_test.go]: We should add an import test step here

File: azurerm/resource_arm_data_factory_integration_runtime_managed_test.go [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMDataFactoryIntegrationRuntimeManagedExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "catalog_info.#", "1"),
+					resource.TestCheckResourceAttrSet(resourceName, "catalog_info.0.server_endpoint"),
+					resource.TestCheckResourceAttr(resourceName, "catalog_info.0.administrator_login", "ssis_catalog_admin"),
+					resource.TestCheckResourceAttr(resourceName, "catalog_info.0.administrator_password", "my-s3cret-p4ssword!"),
+					resource.TestCheckResourceAttr(resourceName, "catalog_info.0.pricing_tier", "Basic"),
+				),
+			},
```
Comment (mbfrahry) [inline azurerm/resource_arm_data_factory_integration_runtime_managed_test.go]: We should add an import test step here

File: website/docs/r/data_factory_integration_runtime_managed.html.markdown [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name of the Managed Integration Runtime. Changing this forces a new resource to be created. Must be globally unique. See the [Microsoft documentation](https://docs.microsoft.com/en-us/azure/data-factory/naming-rules) for all restrictions.
+
+* `data_factory_name` - (Required) Specifies the name of the Data Factory the Managed Integration Runtime belongs to. Changing this forces a new resource to be created.
+
+* `resource_group_name` - (Required) The name of the resource group in which to create the Managed Integration Runtime.
```
Comment (mbfrahry) [inline website/docs/r/data_factory_integration_runtime_managed.html.markdown]:
```suggestion
* `resource_group_name` - (Required) The name of the resource group in which to create the Managed Integration Runtime. Changing this forces a new resource to be created.
```

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @r0bnet. These changes look great! There are some missed errors and checks that we should do and I've called those out below.

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go (hunk 1 above)
Comment (r0bnet) [reply azurerm/resource_arm_data_factory_integration_runtime_managed.go]: Fixed

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go (hunk 3 above)
Comment (r0bnet) [reply azurerm/resource_arm_data_factory_integration_runtime_managed.go]: Fixed

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go (hunk 4 above)
Comment (r0bnet) [reply azurerm/resource_arm_data_factory_integration_runtime_managed.go]: Fixed

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go (hunk 5 above)
Comment (r0bnet) [reply azurerm/resource_arm_data_factory_integration_runtime_managed.go]: Fixed

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go (hunk 7 above)
Comment (r0bnet) [reply azurerm/resource_arm_data_factory_integration_runtime_managed.go]: Fixed

File: azurerm/resource_arm_data_factory_integration_runtime_managed.go (hunk 8 above)
Comment (r0bnet) [reply azurerm/resource_arm_data_factory_integration_runtime_managed.go]: Fixed

File: azurerm/resource_arm_data_factory_integration_runtime_managed_test.go (hunk 9 above)
Comment (r0bnet) [reply azurerm/resource_arm_data_factory_integration_runtime_managed_test.go]: Added import step

**Comment by r0bnet** [comment]
> Hey @r0bnet. These changes look great! There are some missed errors and checks that we should do and I've called those out below.

Fixed most of the stuff except two things: no import steps for catalog info and custom setup script tests as password or sas token is not returned via API so it differs from what is in the state.

**Comment by mbfrahry** [comment]
You can throw in an ImportStateVerifyIgnore to ignore the fields that aren't returned from the API (i.e. `ImportStateVerifyIgnore: []string{"catalog_info.0.administrator_password"},`).

**Comment by r0bnet** [comment]
> You can throw in an ImportStateVerifyIgnore to ignore the fields that aren't returned from the API (i.e. `ImportStateVerifyIgnore: []string{"catalog_info.0.administrator_password"},`).

Ah okay. Thanks. Fixed that for the two tests and it ran the tests locally. Seems to be fine now.

**Review by mbfrahry — APPROVED** [review]
Thanks for fixing this up @r0bnet. Apologies for the rebasing you had to do

**Comment by muralidar44** [comment]
@r0bnet , I am looking for terraform code to create "Datafactory Azure-SSIS Integration runtime" , could you please help, where to get it.

**Comment by r0bnet** [comment]
Hey @muralidar44, 

sure. Looking at the [documentation](https://www.terraform.io/docs/providers/azurerm/r/data_factory_integration_runtime_managed.html) there is already an up&running example existing.

**Pull request merged**